diff -ur a/Orchard.Framework.csproj b/Orchard.Framework.csproj
--- a/Orchard.Framework.csproj	2017-11-02 07:20:31.509081688 +0300
+++ b/Orchard.Framework.csproj	2017-11-02 07:21:33.298990030 +0300
@@ -763,7 +763,6 @@
     <Compile Include="Data\Providers\AbstractDataServicesProvider.cs" />
     <Compile Include="Data\Providers\DataServicesProviderFactory.cs" />
     <Compile Include="Data\Providers\IDataServicesProviderFactory.cs" />
-    <Compile Include="Data\Providers\SqlServerDataServicesProvider.cs" />
     <Compile Include="Data\Providers\MySqlDataServicesProvider.cs" />
     <Compile Include="Data\Conventions\StringLengthConvention.cs" />
     <Compile Include="Data\SessionFactoryHolder.cs" />
