--- a/src/ScriptEngine.HostedScript/ScriptEngine.HostedScript.csproj
+++ b/src/ScriptEngine.HostedScript/ScriptEngine.HostedScript.csproj
@@ -84,7 +84,9 @@
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Ionic.Zip.Reduced" />
+    <Reference Include="Ionic.Zip.Reduced">
+        <HintPath>/usr/share/dev-dotnet/dotnetzip-semverd-1/Ionic.Zip.Reduced.dll</HintPath>
+    </Reference>
     <Reference Include="Newtonsoft.Json">
       <Package>monodevelop</Package>
     </Reference>
