--- a/Source/MSBuild.Community.Tasks/Zip.cs
+++ b/Source/MSBuild.Community.Tasks/Zip.cs
@@ -258,10 +258,10 @@
 
                     if (string.Equals(Encryption, "PkzipWeak", StringComparison.OrdinalIgnoreCase))
                         zip.Encryption = EncryptionAlgorithm.PkzipWeak;
-                    else if (string.Equals(Encryption, "WinZipAes128", StringComparison.OrdinalIgnoreCase))
-                        zip.Encryption = EncryptionAlgorithm.WinZipAes128;
-                    else if (string.Equals(Encryption, "WinZipAes256", StringComparison.OrdinalIgnoreCase))
-                        zip.Encryption = EncryptionAlgorithm.WinZipAes256;
+//                    else if (string.Equals(Encryption, "WinZipAes128", StringComparison.OrdinalIgnoreCase))
+//                        zip.Encryption = EncryptionAlgorithm.WinZipAes128;
+//                    else if (string.Equals(Encryption, "WinZipAes256", StringComparison.OrdinalIgnoreCase))
+//                        zip.Encryption = EncryptionAlgorithm.WinZipAes256;
                     else
                         zip.Encryption = EncryptionAlgorithm.None;
 
