# HG changeset patch
# Parent e487174a0d5a87dc866984adea75ed1cef211d49

diff --git a/configure.in b/configure.in
--- a/configure.in
+++ b/configure.in
@@ -6381,7 +6381,7 @@
             #include "sqlite3.h"
 
             int main(int argc, char **argv){
-              return !sqlite3_compileoption_used("SQLITE_SECURE_DELETE=1");
+              return !sqlite3_compileoption_used("SQLITE_SECURE_DELETE");
             }],
             ac_cv_sqlite_secure_delete=yes,
             ac_cv_sqlite_secure_delete=no,
@@ -6439,7 +6439,7 @@
             #include "sqlite3.h"
 
             int main(int argc, char **argv){
-              return !sqlite3_compileoption_used("SQLITE_ENABLE_FTS3=1");
+              return !sqlite3_compileoption_used("SQLITE_ENABLE_FTS3");
             }],
             ac_cv_sqlite_enable_fts3=yes,
             ac_cv_sqlite_enable_fts3=no,

