--- config/rules.mk.orig	2010-03-14 01:07:46.000000000 +0800
+++ config/rules.mk	2010-03-14 01:07:51.000000000 +0800
@@ -617,8 +617,8 @@
 ifeq ($(OS_ARCH),Linux)
 ifneq (,$(filter mips mipsel,$(OS_TEST)))
 ifeq ($(MODULE),layout)
-OS_CFLAGS += -Wa,-xgot
-OS_CXXFLAGS += -Wa,-xgot
+OS_CFLAGS += -mxgot
+OS_CXXFLAGS += -mxgot
 endif
 endif
 endif
--- js/src/config/rules.mk.orig	2010-03-14 01:06:53.000000000 +0800
+++ js/src/config/rules.mk	2010-03-14 01:07:11.000000000 +0800
@@ -617,8 +617,8 @@
 ifeq ($(OS_ARCH),Linux)
 ifneq (,$(filter mips mipsel,$(OS_TEST)))
 ifeq ($(MODULE),layout)
-OS_CFLAGS += -Wa,-xgot
-OS_CXXFLAGS += -Wa,-xgot
+OS_CFLAGS += -mxgot
+OS_CXXFLAGS += -mxgot
 endif
 endif
 endif
