diff -u -r libgdiplus-4.2/src/text-cairo.c libgdiplus-4.2-patched/src/text-cairo.c
--- libgdiplus-4.2/src/text-cairo.c	2014-08-27 20:45:05.000000000 +0400
+++ libgdiplus-4.2-patched/src/text-cairo.c	2017-12-06 15:40:37.904472973 +0300
@@ -27,7 +27,7 @@
 
 #ifndef USE_PANGO_RENDERING
 
-#include <ctype.h>
+#include <wctype.h>
 
 #include "text-cairo-private.h"
 #include "graphics-private.h"
@@ -237,7 +237,7 @@
 
 	/* unless specified we don't consider the trailing spaces, unless there is just one space (#80680) */
 	if ((format->formatFlags & StringFormatFlagsMeasureTrailingSpaces) == 0) {
-		while ((StringLen > 0) && (isspace ((int) ((unsigned char) *(Src + StringLen - 1)))))
+		while ( (StringLen > 0) && iswspace (Src[StringLen - 1]) )
 			StringLen--;
 		if (StringLen == 0)
 			StringLen = 1;
