--- a/Makefile.am
+++ b/Makefile.am
@@ -1,7 +1,7 @@
 ACLOCAL_AMFLAGS = -I m4
 
-SUBDIRS = src tests
-DIST_SUBDIRS = src tests
+SUBDIRS = src
+DIST_SUBDIRS = src
 
 pkgconfigdir = $(libdir)/pkgconfig
 
@@ -10,10 +10,3 @@
 DISTCLEANFILES= libgdiplus.pc
 
 EXTRA_DIST = libgdiplus.pc.in README.md LICENSE libgdiplus.sln winconfig.h.in update_submodules.sh external/googletest
-
-all: update_submodules
-
-update_submodules:
-	@cd $(top_srcdir) && ./update_submodules.sh
-
-.PHONY: update_submodules
